redis/commands.cc

```diff
@@ -170,22 +170,32 @@ future<redis_message> set(service::storage_proxy& proxy, request& req, redis::re
             }
         }
     }
-    return redis::write_strings(proxy, options, std::move(req._args[0]), std::move(req._args[1]), ttl, permit).then([] {
-        return redis_message::ok();
+    return redis::write_strings(proxy, options, std::move(req._args[0]), std::move(req._args[1]), ttl, permit, ssg).then([] (bool is_applied) {
+        if (is_applied) {
+            return redis_message::ok();
+        }else{
+            // XXX: should retry LWT, instead of returning error
```

Comment:
OH, I think I'm starting to understand what you wanted to do here. It is still wrong, but perhaps now I can explain what is wrong:

I guess you are preparing to have similar code also for "INCR" (maybe you should have started with that operation). To implement INCR with CQL's, you would indeed need to use a loop. BUT, you are not using LWT - you are using the underlying `cas()` feature, and it is more powerful, and doesn't require you to do a loop (it does it for you): Let me explain:

Let's say want a cas_request object that implements INCR. Its `apply()` will take `qr` - the old value of the item, increment it by one, and return that as the mutation. That's it!! You may be wondering how this works in practice. Well, what happens is that `cas()` (thanks to @gleb-cloudius and @kostja for this useful abstraction), serializes the read and the write as one unit. In some high-contention cases (I actually tested those in Alternator), after one node read-and-write it will discover that it actually lost its proposal, and its write never happened - so in this case `cas()` itself will retry - it will read the new value again, and then call `apply()` again. The caller never knows, and doesn't need to retry the LWT.

This actually surprised me to at first. I didn't realize that a cas_request object may have its `apply()` function called more than once - each time with a different timestamp. I had a bug in my first implementation in Alternator where `apply()` assumed it can only be used once, and `std::move()`ed some values stored in the request object. But this was wrong - `apply()` may be called multiple time on the same request object.

Reply:
Reverted the code that write_strings() to return future<bool>.